crates/weaver_cache/src/registry_path.rs

```diff
@@ -0,0 +1,273 @@
+        sub_folder: Option<String>,
+    },
+}
+
+/// Implement the `FromStr` trait for `RegistryPath`, so that it can be used as
+/// a command-line argument.
+impl FromStr for RegistryPath {
+    type Err = Error;
+
+    /// Parse a string into a `RegistryPath`.
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let captures = REGISTRY_REGEX
+            .captures(s)
+            .ok_or(Error::InvalidRegistryPath {
+                path: s.to_owned(),
+                error: "Invalid registry path".to_owned(),
+            })?;
+        let source = captures
+            .name("source")
+            .ok_or(Error::InvalidRegistryPath {
+                path: s.to_owned(),
+                error: "Invalid registry path. No local path or URL found".to_owned(),
+            })?
+            .as_str();
+        let tag = captures.name("tag").map(|m| m.as_str().to_owned());
+        let sub_folder = captures.name("sub_folder").map(|m| m.as_str().to_owned());
+
+        if source.starts_with("http://") || source.starts_with("https://") {
+            if source.ends_with(".zip") || source.ends_with(".tar.gz") {
+                Ok(Self::RemoteArchive {
+                    url: source.to_owned(),
+                    sub_folder,
+                })
+            } else {
+                Ok(Self::GitRepo {
+                    url: source.to_owned(),
+                    tag,
+                    sub_folder,
+                })
+            }
+        } else if source.ends_with(".zip") || source.ends_with(".tar.gz") {
+            Ok(Self::LocalArchive {
+                path: source.to_owned(),
+                sub_folder,
+            })
+        } else {
+            Ok(Self::LocalFolder {
+                path: source.to_owned(),
+            })
+        }
+    }
+}
+
+/// Implement the `Display` trait for `RegistryPath`, so that it can be printed
+/// to the console.
+impl Display for RegistryPath {
+    /// Format the `RegistryPath` as a string.
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            RegistryPath::LocalFolder { path } => write!(f, "{}", path),
+            RegistryPath::LocalArchive { path, sub_folder } => {
+                if let Some(sub_folder) = sub_folder {
+                    write!(f, "{}[{}]", path, sub_folder)
+                } else {
+                    write!(f, "{}", path)
+                }
+            }
+            RegistryPath::RemoteArchive { url, sub_folder } => {
+                if let Some(sub_folder) = sub_folder {
+                    write!(f, "{}[{}]", url, sub_folder)
+                } else {
+                    write!(f, "{}", url)
+                }
+            }
+            RegistryPath::GitRepo {
+                url,
+                tag,
+                sub_folder,
+            } => {
+                let mut registry_path = url.clone();
```

Comment:
(extra) Nit: I find this more readable with a match, so you can see the format permutations and no `.clone()` call.

```rust
match (tag, sub_folder) {
  (Some(tag), Some(folder)) => write!(f, "{}@{}[{}]", url, tag, folder),
  (Some(tag), None) => write!(f, "{}@{}", url, tag),
  (None, Some(folder)) => write!(f, "{}[{}]", url, folder),
  (None, None) => write!(f, "{}", url),
}
```

Reply:
Integrated. I like it.